# Segmenting and Clustering Neighbourhoods in Toronto
### First of all, let's import the pandas library and read the html of the Wikipedia url.

In [38]:
import pandas as pd

pd.read_html('https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&oldid=1011037969')

dfs = pd.read_html('https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&oldid=1011037969', header=0)

### By using [0] as our header, we can now determine a dataframe called 'df'.

In [39]:
df = dfs[0]
df

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

As we can see above, there's a lot of 'Not assigned' values. We need to fix that!

### To fix these values, we will simply create a filtered dataframe which will not have any 'Not assigned' values.

In [40]:
df_filtered = df[df['Borough'] != 'Not assigned'] 

In [41]:
df_filtered

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

It is clear that the neighbourhoods are already grouped by the Postal Code, so there isn't any 'Not assigned' values in the 'Neighbourhood' column.

### Now we can finally check the dataframe shape by using the .shape function!

In [42]:
df_filtered.shape

(103, 3)

So the new dataframe has 103 rows and 3 columns!

### Let's import the Geospatial Coordinates data.
We also need to check if it matches the same number of rows and columns from our filtered dataframe.

In [43]:
import os, types
import pandas as pd
from botocore.client import Config
import ibm_boto3

def __iter__(self): return 0

# @hidden_cell


df_data_1 = pd.read_csv(body)
df_data_1.shape

(103, 3)

As shown by the .shape function, the Geospatial_Coordinates.csv has 103 rows and 3 columns too! It matches the filtered dataframe .shape function.

### We can proceed to merge the two dataframes.
Let's use the .merge() function!

In [44]:
new_df = df_filtered.merge(df_data_1,left_on='Postal Code',right_on='Postal Code')

In [45]:
new_df.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

There it is! The merged table is ready for use!